packages/integration-objects/srv/integration-objects/simple-exchange-rate-type-service.js

```diff
@@ -0,0 +1,210 @@
+    }
+  }
+
+  async function beforeUpdate(req) {
+    try {
+      await validatePayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validatePayload(req) {
+    validateTenantIdInPayload(req);
+    validateFieldsForExchangeRateTypes(req.data);
+    await checkUniquenessForPrimaryKeys(req);
+  }
+
+  function beforeDelete(req) {
+    try {
+      validateDelete(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeCreateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeUpdateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validateDescPayload(req) {
+    validateFieldsForExchangeRateTypeDescription(req.data, req.params);
+    await checkForValidityOfKey(req, ExchangeRateTypes);
+    await checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req);
+  }
+
+  async function checkUniquenessForPrimaryKeys(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingRateTypes(ExchangeRateTypes, record, req);
+    checkForDuplicateRateTypeObjects(affectedRows, req);
+  }
+
+  async function checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingRateTypeDesc(ExchangeRateTypeDescriptions, record, req);
+    checkForDuplicateRateTypeObjects(affectedRows, req);
+  }
+
+  async function beforeActiveDeleteForExchangeRateTypeDescription(req) {
+    const record = req.data;
+    let id;
+    const affectedRows = await fetchExistingTexts(ExchangeRateTypeDescriptions, record, req);
+    if (affectedRows.length) {
+      id = affectedRows[0].ID_texts;
+    }
+    await validateIdInTenant(id, ExchangeRateTypes, req, affectedRows);
+  }
+
+  async function writeAuditLogForRateTypes(req, next) {
+    let currentValues;
+    const data = req.data;
+    const type = 'exchange-rate-type';
+    const auditParams = {
+      exchangeRateType: data.exchangeRateType,
+      exchangeRateTypeDescription: data.exchangeRateTypeDescription,
+      referenceCurrencyThreeLetterISOCode: data.referenceCurrencyThreeLetterISOCode
+    };
+
+    let currentValuesIsNotEmpty = true;
+    if (Constants.CREATE_EVENT !== req.event) {
+      currentValues = await fetchCurrentAttributeValues(req, req.user.tenant, ExchangeRateTypes);
+      currentValuesIsNotEmpty = !util.isNullish(currentValues);
+    }
+
+    if (currentValuesIsNotEmpty) {
+      await writeAuditLog(req, auditParams, currentValues, type);
+    }
+    await next();
+  }
+};
+
+async function validateIdInTenant(id, ExchangeRateTypes, req, affectedRows) {
```

Comment:
This seems a bit weird to me. The first parameter, that we want to check is optional. But if it is not there, the function doesn't make sense, because I want to validate the id. The id should not be allowed to be `undefined` and instead the function should not be called if there is no id to validate (see above).

Reply:
Updated this.

Reply:
👍